Agent/Services/AgentHubConnection.cs

```diff
@@ -172,373 +192,459 @@ public async Task Connect()
+            EnsureHubConnection();
+
+            if (!_isServerVerified)
             {
+                _logger.LogWarning(
+                    "Command attempted before server was verified.  Shell: {shell}.  Command: {command}.  Sender: {senderConnectionID}",
+                    shell,
+                    command,
+                    senderConnectionId);
                 return;
             }
 
-            if (string.IsNullOrWhiteSpace(_connectionInfo.OrganizationID))
+            await _scriptExecutor.RunCommandFromTerminal(
+                    shell,
+                    command,
+                    authToken,
+                    senderUsername,
+                    senderConnectionId,
+                    ScriptInputType.Terminal,
+                    TimeSpan.FromSeconds(30),
+                    _hubConnection)
+                .ConfigureAwait(false);
+        }
+        catch (Exception ex)
+        {
+            _logger.LogError(ex, "Error while executing command.");
+        }
+    }
+
+    public async Task ExecuteCommandFromApi(ScriptingShell shell, string authToken, string requestID, string command, string senderUsername)
+    {
+        try
+        {
+            EnsureHubConnection();
+
+            if (!_isServerVerified)
             {
-                _logger.LogError("Organization ID is not set.  Please set it in the config file.");
+                _logger.LogWarning(
+                    "Command attempted before server was verified.  Shell: {shell}.  Command: {command}.  Sender: {senderUsername}",
+                    shell,
+                    command,
+                    senderUsername);
                 return;
             }
 
-            var currentInfo = await _deviceInfoService.CreateDevice(_connectionInfo.DeviceID, _connectionInfo.OrganizationID);
-            await _hubConnection.SendAsync("DeviceHeartbeat", currentInfo);
+            await _scriptExecutor
+                .RunCommandFromApi(shell, requestID, command, senderUsername, authToken, _hubConnection)
+                .ConfigureAwait(false);
         }
         catch (Exception ex)
         {
-            _logger.LogWarning(ex, "Error while sending heartbeat.");
+            _logger.LogError(ex, "Error while executing command from API.");
         }
     }
 
-    private async Task<bool> CheckForServerMigration()
+    public async Task GetLogs(string senderConnectionId)
     {
-        if (_connectionInfo is null || _hubConnection is null)
+        try
         {
-            return false;
-        }
+            EnsureHubConnection();
 
-        var serverUrl = await _hubConnection.InvokeAsync<string>("GetServerUrl");
+            if (!await _fileLogsManager.AnyLogsExist(_appLifetime.ApplicationStopping))
+            {
+                var message = "There are no log entries written.";
+                await _hubConnection.InvokeAsync("SendLogs", message, senderConnectionId).ConfigureAwait(false);
```

Comment:
I wonder if we'd be better off using payloads / DTOs in these hub methods, to make changes to the api easier to implement across server and client. E.g. 

```c#
_hubConnection.InvokeAsync("SendLogs", new SendLogsPayload(message, senderConnectionId))
```

so that if we later want to add properties to the payload, the client still accepts in the same number of parameters and has a good shot of being able to work even if it hasn't been updated.

I realize this PR is more of a restructuring than it is writing all this code afresh, though. So perhaps this is more of a suggestion for future refactoring